### DISCUSSION TITLE
.NET Standard support

### DISCUSSION THREAD

**Discussion by YoshiRulz** [amerkoleci/Vortice.Windows#416] [General]
The `netstandard2.0` target was dropped in https://github.com/amerkoleci/Vortice.Windows/commit/0445f50f26b471b9d9147cf98c2da3e99aa23836 but it's not clear why. Is it possible to add it back? If not, could you release a 2.4.3 with updated NuGet dependencies?

**Comment by amerkoleci** [comment]
Hi,
That won't be possible, as I'll need to keep lot of workarounds to support netstandard2.0, if you find clear way feel free to send PR and I can think about it.

Thanks!

**Reply by YoshiRulz** [reply]
I'm almost done with the modifications to Vortice.Windows—it's compiling, and the total changeset (edit: unlinked as branch was deleted) is fairly small. But to even start it was necessary to temporarily downgrade Vortice.Mathematics to the last version that targeted .NET Standard. I'm now looking at adding the target back to the current Vortice.Mathematics, and I'm guessing that restoring the compile-time guards from https://github.com/amerkoleci/Vortice.Mathematics/commit/3730dd76a58e2a3e500fea2e1b93791d37450f66 will be sufficient.

edit: I see you're in the middle of some refactors, so I'll stop there for now. I hope you end up with less duplicate code and are able to look into shimming `Vector128<T>` directly (there doesn't seem to be an existing polyfill for it, but I'd be willing to help create one).

**Reply by BradChase2011** [reply]
YoshiRulz, I'm extremely interested in this if you could keep us updated, I would be in debt.  It would be nice to use the latest dx bits that don't exist in SharpDX.  It would let us test some more stuff.

**Reply by YoshiRulz** [reply]
Revisited `Vortice.Mathematics` this week. I did end up writing a polyfill for `Vector128<T>`, and that got me most of the way to .NET Standard support. You can see my changeset [here](https://github.com/YoshiRulz/Vortice.Mathematics/compare/main...netstandard). I'm sure maintaining the polyfill in this repo doesn't sound appealing, so I'd like to distribute it on NuGet myself, though I haven't looked much into that, and such a package should probably include the whole API surface whereas I've only done the small fraction that `Vortice.Mathematics` uses.

**Comment by YoshiRulz** [comment]
I tried building a few commits but they all failed with something like this:
```
$> git checkout fd75e80
$> dotnet pack Vortice.Windows.sln --configuration Release -p:Packing=true -p:EnableWindowsTargeting=true
MSBuild version 17.8.0+6cdef4241 for .NET
  Determining projects to restore...
  Restored /home/yoshi/Documents/Vortice.Windows/src/tests/Vortice.Dxc.Test/Vortice.Dxc.Test.csproj (in 461 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/tests/Vortice.DXGI.Tests/Vortice.DXGI.Tests.csproj (in 461 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/HelloDirectStorage/HelloDirectStorage.csproj (in 464 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.XAudio2/Vortice.XAudio2.csproj (in 463 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/XInput/HelloXInput.csproj (in 6 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/MediaFoundation/EnumerateDevices/EnumerateDevices.csproj (in 6 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/MediaFoundation/AudioPlay/AudioPlay.csproj (in 9 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/HelloXAudio2/HelloXAudio2.csproj (in 8 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/tests/Vortice.D3DCompiler.Test/Vortice.D3DCompiler.Test.csproj (in 28 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/HelloDirectML/HelloDirectML.csproj (in 25 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/HelloDirectInput/HelloDirectInput.csproj (in 22 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/Vortice.SampleFramework/Vortice.SampleFramework.csproj (in 64 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/HelloDirect3D12Raytracing/HelloDirect3D12Raytracing.csproj (in 13 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/HelloDirect3D12/HelloDirect3D12.csproj (in 16 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/AdvancedTextRenderingApp/AdvancedTextRenderingApp.csproj (in 13 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.XInput/Vortice.XInput.csproj (in 5 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/samples/HelloDirect3D11/HelloDirect3D11.csproj (in 24 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.DirectML/Vortice.DirectML.csproj (in 21 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.UIAnimation/Vortice.UIAnimation.csproj (in 14 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.Dxc/Vortice.Dxc.csproj (in 5 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.MediaFoundation/Vortice.MediaFoundation.csproj (in 8 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.DirectX/Vortice.DirectX.csproj (in 20 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.DirectStorage/Vortice.DirectStorage.csproj (in 9 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.DirectSound/Vortice.DirectSound.csproj (in 7 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.DirectInput/Vortice.DirectInput.csproj (in 21 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.Direct3D9/Vortice.Direct3D9.csproj (in 7 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.DirectComposition/Vortice.DirectComposition.csproj (in 9 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.Direct3D12/Vortice.Direct3D12.csproj (in 11 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.Direct2D1/Vortice.Direct2D1.csproj (in 13 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.Direct3D11/Vortice.Direct3D11.csproj (in 21 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.DXGI/Vortice.DXGI.csproj (in 11 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.D3DCompiler/Vortice.D3DCompiler.csproj (in 12 ms).
  Restored /home/yoshi/Documents/Vortice.Windows/src/Vortice.WinUI/Vortice.WinUI.csproj (in 153 ms).
/home/yoshi/.nuget/packages/sharpgentools.sdk/2.1.2-beta/Sdk/Sdk.targets(191,5): error MSB4044: The "SharpGenTask" task was not given a value for the required parameter "CastXmlExecutable". [/home/yoshi/Documents/Vortice.Windows/src/Vortice.Dxc/Vortice.Dxc.csproj]
/home/yoshi/.nuget/packages/sharpgentools.sdk/2.1.2-beta/Sdk/Sdk.targets(191,5): error MSB4044: The "SharpGenTask" task was not given a value for the required parameter "CastXmlExecutable". [/home/yoshi/Documents/Vortice.Windows/src/Vortice.DirectX/Vortice.DirectX.csproj]
  Vortice.XInput -> /home/yoshi/Documents/Vortice.Windows/bin/Vortice.XInput/Release/net7.0/Vortice.XInput.dll
  Vortice.XInput -> /home/yoshi/Documents/Vortice.Windows/bin/Vortice.XInput/Release/net8.0/Vortice.XInput.dll
/home/yoshi/.nuget/packages/sharpgentools.sdk/2.1.2-beta/Sdk/Sdk.targets(191,5): error MSB4044: The "SharpGenTask" task was not given a value for the required parameter "CastXmlExecutable". [/home/yoshi/Documents/Vortice.Windows/src/Vortice.Direct3D9/Vortice.Direct3D9.csproj]
/home/yoshi/.nuget/packages/sharpgentools.sdk/2.1.2-beta/Sdk/Sdk.targets(191,5): error MSB4044: The "SharpGenTask" task was not given a value for the required parameter "CastXmlExecutable". [/home/yoshi/Documents/Vortice.Windows/src/Vortice.DirectX/Vortice.DirectX.csproj::TargetFramework=net7.0]
/home/yoshi/.nuget/packages/sharpgentools.sdk/2.1.2-beta/Sdk/Sdk.targets(191,5): error MSB4044: The "SharpGenTask" task was not given a value for the required parameter "CastXmlExecutable". [/home/yoshi/Documents/Vortice.Windows/src/Vortice.UIAnimation/Vortice.UIAnimation.csproj]
  The package Vortice.XInput.3.4.1-beta is missing a readme. Go to https://aka.ms/nuget/authoring-best-practices/readme to learn why package readmes are important.
  Successfully created package '/home/yoshi/Documents/Vortice.Windows/artifacts/Vortice.XInput.3.4.1-beta.nupkg'.
```
I'm guessing it relates to SharpGenTools/SharpGenTools#111.

My other suggestion remains.

**Reply by BradChase2011** [reply]
I presume you are hitting UWP or XBox?  We have run into the same limitations.  We made an entire conversion over to the old SharpDX, but they are missing alot of stuff.  For us, little things like being able to grab the audio device id's off the gamepads dont exist and trying to build SharpDX and update it seems to be a massive burden.  Vortice seems to have alot of really nice extensions in use now for vectors which speed things up massively so I can understand how hard it would be to go backwards there...  Just thinking out loud but last I looked it seemed most of the math stuff is either available from microsoft or can be copied out.  Maybe some sort of false vector support for netstandard2 could be added in.  I was actually going to go down that route but ended up coming to the idea that going UWP and XBox community is really a dead end.  We REALLY want to have a community release of our engine but Microsoft seems to want to fight anyone wanting to go that route.

**Reply by YoshiRulz** [reply]
Actually ours is a WinForms desktop app—.NET Framework WinForms that is. Obviously we'll have to switch UI frameworks in order to move to .NET 8, but we're trying to separate concerns and leave *only* UI changes for that final PR.

For game dev, can't you use Veldrid + Dear ImGui or something?

**Reply by amerkoleci** [reply]
As I said, netstandard2.0 isn't coming back to be supported, try other ways like updating to net7 or net8, or fork the project and add it back (if you want to maintain separate fork OFC).

**Reply by BradChase2011** [reply]
Why can't you move to .net 8 with WinForms?  Seems that would be the path of least resistance, not to mention a hell of a lot faster.  As for us, we have our own engine and multiple targets OSs, we wanted to hit Xbox community but like I said I doubt Microsoft will even have that around much longer considering they don't even bother updating their support docs any longer and UWP has been deprecated.  I can't think of any other reason netstandard2 would even be needed outside of that.  Is there a reason you haven't moved to .net8 with WinForms?

**Reply by YoshiRulz** [reply]
The app runs on both Windows and Linux, and formerly macOS too, which we may yet return to. The [new WinForms](https://github.com/dotnet/winforms) is Windows-only.

In hindsight it does seem like .NET Standard was a failed experiment. I expect that once we've moved to .NET 8, all our internal-use libraries which currently target .NET Standard will be retargeted as well.

**Reply by BradChase2011** [reply]
Oh wow, I had no idea the old WinForms could run outside of windows.  Ya learn something new every day :).

**Comment by MarcioAB** [comment]
Using Net 8 with Windows Forms here.
Not sure if I understood the issue.

**Reply by YoshiRulz** [reply]
Assuming this is a reply to https://github.com/amerkoleci/Vortice.Windows/discussions/416#discussioncomment-7953164: The [WinForms you're familiar with](https://github.com/dotnet/winforms) is Windows-only. The old one was cross-platform if you stuck to the beaten track.
